### PR TITLE
[Merged by Bors] - tortoise: store ballots separately from layers

### LINKED ISSUE #3779: tortoise: stored ballots are not properly evicted
image

### PR DESCRIPTION
**Pull request by dshulyak**
close: https://github.com/spacemeshos/go-spacemesh/issues/3779

each ballot stores votes on X previous layers. ballots referenced in the layer store votes for another X layers and so on. this is clearly recursive and can't be used like this.

### CODE REVIEW

File: tortoise/state.go [hunk 1]
```diff
@@ -62,8 +62,9 @@ type (
 			min, max types.LayerID
 		}
 
-		epochs map[types.EpochID]*epochInfo
-		layers map[types.LayerID]*layerInfo
+		epochs  map[types.EpochID]*epochInfo
+		layers  map[types.LayerID]*layerInfo
+		ballots map[types.LayerID][]*ballotInfo
```
Comment (countvonzero) [inline tortoise/state.go:71]: it's definitely worth adding a comments about this recursive behavior.
i think it'll be very tempting for future dev to do the same thing without realizing the subtleties.

**Review by countvonzero — APPROVED** [review]

File: tortoise/state.go (hunk 1 above)
Comment (dshulyak) [reply tortoise/state.go:71]: definitely, added a comment

**Comment by dshulyak** [comment]
memory consistently around 300-400 mb. seems a bit high, but that can be explored separately

image

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Pull request closed without merging**